lib/src/configuration.dart

```diff
@@ -395,21 +418,315 @@ class RealmSchema extends Iterable<SchemaObject> {
+  /// The callback that will be executed just before the client reset happens.
+  BeforeResetCallback? get onBeforeReset => _onBeforeReset;
+
+  /// The callback that will be executed just after the client reset happens if the local changes
+  /// needed to be discarded.
+  AfterResetCallback? get onAfterDiscard => _onAfterDiscard;
+
+  /// The callback that will be executed just after the client reset happens if the local changes
+  /// were successfully recovered.
+  AfterResetCallback? get onAfterRecovery => _onAfterRecovery;
+
+  /// Creates an instance of `RecoverOrDiscardUnsyncedChangesHandler`.
+  ///
+  /// This strategy will recover automatically any unsynchronized changes. If the recovery fails this strategy fallsback to the discard unsynced one.
+  /// The automatic recovery mechanism creates write transactions meaning that all the changes that take place
+  /// are properly propagated through the standard Realm's change notifications.
+  /// This strategy supplies four callbacks: [onBeforeReset], [onAfterRecovery], [onAfterDiscard] and [onManualResetFallback].
+  /// [onBeforeReset] is invoked just before the client reset happens.
+  /// [onAfterRecovery] is invoke if and only if an automatic client reset succeeded. The callback is never called
+  /// if the automatic client reset fails.
+  /// [onAfterDiscard] is invoked if and only if an automatic client reset failed and instead the discard unsynced one succeded.
+  /// The callback is never called if the discard unsynced client reset fails.
+  /// [onManualResetFallback] is invoked whenever an error occurs in either of the recovery stragegies and the system needs to fallback to a manual mode.
+  const RecoverOrDiscardUnsyncedChangesHandler(
+      {BeforeResetCallback? onBeforeReset, AfterResetCallback? onAfterRecovery, AfterResetCallback? onAfterDiscard, ClientResetCallback? onManualResetFallback})
+      : super._(ClientResyncModeInternal.recoverOrDiscard, onManualResetFallback,
+            onBeforeReset: onBeforeReset, onAfterDiscard: onAfterDiscard, onAfterRecovery: onAfterRecovery);
+}
 
 /// @nodoc
-extension SchemaObjectInternal on SchemaObject {
-  bool get isGenericRealmObject => type == RealmObject || type == EmbeddedObject || type == RealmObjectBase;
+extension ClientResetHandlerInternal on ClientResetHandler {
+  ClientResyncModeInternal get clientResyncMode => _mode;
+
+  BeforeResetCallback? get onBeforeReset => _onBeforeReset;
+  AfterResetCallback? get onAfterDiscard => _onAfterDiscard;
+  AfterResetCallback? get onAfterRecovery => _onAfterRecovery;
+}
+
+/// Enum describing what should happen in case of a Client Resync.
+/// @nodoc
+enum ClientResyncModeInternal {
+  manual,
+  discardLocal,
+  recover,
+  recoverOrDiscard,
+}
+
+/// An error type that describes a client reset error condition.
+/// {@category Sync}
+class ClientResetError extends SyncError {
+  /// If true the received error is fatal.
+  final bool isFatal = true;
+  final Configuration? config;
```

Comment:
why do we have the config here or why is it public?

Reply:
We use it for `resetRealm` method. I made it private.